src/lib.rs

```diff
@@ -2268,6 +2270,40 @@ pub trait Itertools : Iterator {
         v.into_iter()
     }
 
+    /// Sort the k smallest elements into a new iterator, in ascending order.
+    ///
+    /// **Note:** This consumes the entire iterator, and returns the result
+    /// as a new iterator that owns its elements.
+    ///
+    /// This is guaranteed to use `k * sizeof(Self::Item) + O(1)` memory
+    /// and `O(n log k)` time, with `n` the number of elements that yielded
+    /// by the iterator.
+    ///
+    /// The sorted iterator, if directly collected to a `Vec`, is converted
+    /// without any extra copying or allocation cost.
+    ///
+    /// ```
+    /// use itertools::Itertools;
+    ///
+    /// // A random permutation of 0..15
+    /// let numbers = vec![6, 9, 1, 14, 0, 4, 8, 7, 11, 2, 10, 3, 13, 12, 5];
+    ///
+    /// let five_smallest = numbers
+    ///     .into_iter()
+    ///     .k_smallest(5);
+    ///
+    /// itertools::assert_equal(five_smallest, 0..5);
+    /// ```
+    #[cfg(feature = "use_std")]
+    fn k_smallest(self, k: usize) -> VecIntoIter<Self::Item>
```

Comment:
Not sure what's `itertools` policy but maybe if the return type was newtype'd as `KSmallestIter` or something, it would allow changing it into the future to something more efficient?

Reply:
That's a good point, I'll address that tomorrow  <3

Reply:
I forgot to mention, but that doesn't seem feasible while keeping the guarantee that `.collect::<Vec<_>>` is “free” (in-place, O(1) time).  The stdlib's vec iterator can do it, but does so using specialisation (which isn't stabilised yet)